##  Importing Libraries to Computer (Pandas, NumPy, and OS)

In [1]:
# Importing libraries to computer
import pandas as pd
import numpy as np
import os

## Importing Products and Orders_Wrangled Datasets

In [2]:
# Creating a data importing shortcut called 'path' for the Instacart folder to make importing data easier and cleaner
path =r'/Users/drewsmith/Desktop/01-2023 Instacart Basket Analysis'

In [3]:
# Utilizing the created 'path' shortcut to import the 'Products' dataset 
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [62]:
# Utilizing the created 'path' shortcut to import the 'Orders_wrangled' dataset
df_ords1 = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

#### *** I changed the name of the 'df_ords' to 'df_ords1' because I found an additional column called 'Unnamed: 0' in the original df_ords dataframe and removed the column through a command but I sent the command twice which gave me an error since it couldn't find the column name (since I removed it). Therefore I changed the dataframe name so I could do the column removal command without an error

### --Test Dataframe!!!

In [5]:
# Creates a test dataframe
df_test=pd.DataFrame()

In [6]:
# Creates a mixed type column
df_test['mix']=['a','b',1,True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Checks for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Converts the columns data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

## Finding Missing Values in Products Dataframe (READING EXERCISES)

In [10]:
# Used to find missing observations or entries in the products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [70]:
# Creates a subset with only rows containing missing values under the 'product_name' column
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Addressing Missing Values in Products Dataframe

In [23]:
# Finds count of rows and columns in dataframe
df_prods.shape

(49693, 5)

In [25]:
# Creates dataframe for non-missing 'product_name' values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [27]:
# Shows the number of rows decreased by 16 since it doesn't include missing values
df_prods_clean.shape

(49677, 5)

## Finding Duplicates in Products Dataframe

In [28]:
# Displays only rows that are duplicates in df_prods_clean subset under a new subset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [31]:
# Shows the number of rows and columns in this dataframe
df_prods_clean.shape

(49677, 5)

In [34]:
# Creates new subset of the 'df_prods_clean' dataframe but with the duplicates dropped
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [35]:
# Shows the number of rows decreased by 5 since there were 5 duplicates dropped
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 Questions/Answers

## 02. (Instructions say) Use describe function for orders dataframe to see if any data looks off or investigated further

In [63]:
# ***Removed the 'Unnamed: 0' column since there was an error from exporting the data in the previous exercise. I had to rename the df_ords dataframe since I accidently confirmed the code twice***
df_ords1 = df_ords1.drop(columns=['Unnamed: 0'])

In [64]:
# Shows descriptive statistics for the 'orders' dataframe
df_ords1.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Answer: I don't see anything that looks off especially for min and max values.

## 02. (Rubric says) Use describe function for products dataframe to see if any data looks off or investigated further

In [100]:
# Shows descriptive statistics for the 'products' dataframe
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### Answer: The prices column has a max price that looks wrong since I doubt any product costs 99999 dollars.

## 03./04. Check for mixed-type data in df_ords1 dataframe

In [65]:
# Checks for mixed types. Will display if each column is mixed or consistent
for col in df_ords1.columns.tolist():
 weird = (df_ords1[[col]].applymap(type) != df_ords1[[col]].iloc[0].apply(type)).any(axis = 1)
 if len (df_ords1[weird]) > 0:
        print (col, ' mixed')
 else: print(col, ' consistent')


order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


### Answer: This is no mixed data found in the df_ords1 dataframe

## 05. Run a check for missing values df_ords1 dataframe

In [66]:
# Used to find missing observations or entries in the dataframe
df_ords1.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [71]:
# Creates a subset with only rows containing missing values under the 'days_since_prior_order' column
df_missing = df_ords1[df_ords1['days_since_prior_order'].isnull() ==True]

In [85]:
df_missing.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
11   2168274        2             1                   2                 11   
26   1374495        3             1                   1                 14   
39   3343014        4             1                   6                 11   
45   2717275        5             1                   3                 12   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN

### Answer: I found the 'days_since_prior_order' column contains 206209 missing values. I believe the missing values could possibly mean that those are new customers which obviously they wont have a prior order.

## 06. Address the missing values using an appropriate method

In [87]:
# Tells how many rows and columns there are
df_ords1.shape

(3421083, 6)

In [90]:
# Creating new subset for dataframe
df_ords_new = df_ords1

In [91]:
# Creates new_customer column and puts true values along with null values in the days_since_prior_order column
df_ords_new['new_customer'] = df_ords1['days_since_prior_order'].isnull() == True

In [93]:
df_ords_new.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   

    days_since_prior_order  new_customer  
0                      NaN          True  
1                     15.0         False  
2                     21.0         False  
3                     29.0         False  
4                     28.0         False  
5                     19.0         False  
6                     20.0         False  
7                     14.0         False  
8                      0.0         False  
9                     30.0         False  
10                    14.0         False  
11                     NaN          True  
12                    10.0         False  
13                     3.0         False  
14                     8.0         False  
15                     8.0         False  
16                    13.0         False  
17                    14.0         False  
18                    27.0         False  
19                     8.0         False

### Answer: I created a new column for the new_customers to see if whether them being a new customer (true) aligns with a missing value under the 'days_since_prior_order' column. The results confirm the assumption about why there are missing values since everytime the new_customer column says 'true', it is also the first time a customer has ordered which shows a missing value under the 'days_since_prior_order' column. Any purchase after for that 'user_id' has a number under the 'days_since_prior_order' column and a corresponding false value under 'new_customer'

## 07./08. Run a check for duplicate values in your df_ords1 data

In [95]:
# Displays only rows that are duplicates in df_ords1 dataframe under a new subset
df_ords1_dups = df_ords1[df_ords1.duplicated()]

In [96]:
df_ords1_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

### Answer: There were no duplicates found

## 09. Export your final, cleaned df_prods and df_ords1 data as “.csv” files in your “Prepared Data” folder and give them appropriate names

In [102]:
df_ords1.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [103]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))